### ISSUE TITLE
Schema editor

### ISSUE THREAD

**Issue by amercader** [frictionlessdata/ckanext-validation#65]
Resurrect @roll 's ancient PR https://github.com/frictionlessdata/ckanext-validation/pull/25 and update it with modern components

TODO expand

**Comment by amercader** [comment]
@roll, any thoughts on tableschema-js vs frictionless-py? (see above)

**Comment by roll** [comment]
Hi @amercader,

I would vote for `frictionless-py` way as `tableschema-js` is more like in maintenance-only mode and a more realistic understanding of the situation is that OFKN will not be able to support it long-term. Actually, it's been already moved to the "Universe" from our core products and was maintained by Datopian (a little bit).

Technically, my suggestion would be:
- creating an endpoint that accepts a url or a file sample and returns a Resource descriptor inferred by `frictionless-py`. 
- on CKAN UI we can act on File Upload change event for it. So the user will get Resource/Schema editor during the main resource creation step

I think a one-step arch is more promising as it might be used later to provide types for Data Pusher / Indexer

PS.
Regarding UI I think we need to wait a few months to have new generation of Frictionless Components released

**Comment by amercader** [comment]
@roll sorry, revisiting this after a while. When you say
> *  creating an endpoint that accepts a url or a file sample and returns a Resource descriptor inferred by frictionless-py.
> *  on CKAN UI we can act on File Upload change event for it. So the user will get Resource/Schema editor during the main resource creation step

do you mean the following:

1. User clicks on "Upload" and selects a file
2. We listen to the file input event, and if it's a suitable file (ie tabular) we do a background HTTP request sending a sample of the file (or all of it if it's small enough) to an endpoint that gets a sample tabular data and outputs a Table Schema descriptor
3. With the returned Table Schema descriptor returned we render the Schema Editor component

So essentially is option 2c: Upload a sample of the file, infer the schema, create the resource (and upload the file)
Conceptually it doesn't seem far from 2b but without the complexities of re-factoring the whole CKAN upload process, so it can be a first initial step.
Any thoughts on how big a sample we should upload to have reliable results?

> Although it needs to be investigated regarding compatibility with Excel/etc files
What do you mean, that for Excel files we might not be able to get a sample?

**Comment by roll** [comment]
@amercader 
Yes, it's a good flow description :+1: By default, frictionless uses quite a minimalistic:
- buffer size for encoding inference - 10 000 bytes
- sample size for dialect/schema inference - 100 lines

In most cases, it works fine, and the user will be able to tweek the results anyway.

Regarding Excel, I think it will require sending the whole file to the server (or reading it client-side) just because of the format structure (ZIP index written in the end). I guess Excel is not so sensitive to the size problem as really big data usually in csv

**Comment by amercader** [comment]
Revised implementation plan after discussion with @aivuk 

1. When the user selects a tabular file, we upload it in the background using a custom endpoint that:
   * Creates a new resource with just the uploaded file
   * Infers the schema using the whole uploaded file 
   * Returns the new resource_id and the inferred schema
2. The user can keep entering the rest of the fields and when we get the inferred schema, we update the UI to show a preview and the schema editor
3. When the user clicks "Save" (or "Save and add another") we call another custom endpoint that calls resource_patch on the previously created resource with the rest of the values sent.
4. If the user clicks "Cancel" (or leaves the page?) we delete the resource and the file
